ipa-core/src/protocol/ipa_prf/shuffle/malicious.rs

```diff
@@ -0,0 +1,319 @@
+use std::{borrow::Borrow, iter};
+
+use futures_util::future::{try_join, try_join3};
+
+use crate::{
+    error::Error,
+    ff::{boolean_array::BooleanArray, Field, Gf32Bit},
+    helpers::{
+        hashing::{compute_hash, Hash},
+        Direction, Role, TotalRecords,
+    },
+    protocol::{
+        basics::malicious_reveal,
+        context::Context,
+        ipa_prf::shuffle::{base::IntermediateShuffleMessages, step::OPRFShuffleStep},
+        RecordId,
+    },
+    secret_sharing::{
+        replicated::{semi_honest::AdditiveShare, ReplicatedSecretSharing},
+        SharedValue, StdArray,
+    },
+};
+
+/// This function verifies the `shuffled_shares` and the `IntermediateShuffleMessages`.
+///
+/// ## Errors
+/// Propagates network errors.
+/// Further, returns an error when messages are inconsistent with the MAC tags.
+async fn verify_shuffle<C: Context, S: BooleanArray>(
+    ctx: C,
+    key_shares: &[AdditiveShare<Gf32Bit>],
+    shuffled_shares: &[AdditiveShare<S>],
+    messages: IntermediateShuffleMessages<S>,
+) -> Result<(), Error> {
+    // reveal keys
+    let k_ctx = ctx
+        .narrow(&OPRFShuffleStep::RevealMACKey)
+        .set_total_records(TotalRecords::specified(key_shares.len())?);
+    let keys = reveal_keys(&k_ctx, key_shares).await?;
+
+    // verify messages and shares
+    match ctx.role() {
+        Role::H1 => h1_verify(ctx, &keys, shuffled_shares, messages.get_x1_or_y1()).await,
+        Role::H2 => h2_verify(ctx, &keys, shuffled_shares, messages.get_x2_or_y2()).await,
+        Role::H3 => {
+            h3_verify(
+                ctx,
+                &keys,
+                shuffled_shares,
+                messages.get_x1_or_y1(),
+                messages.get_x2_or_y2(),
+            )
+            .await
+        }
+    }
+}
+
+/// This is the verification function run by `H1`.
+/// `H1` computes the hash for `x1` and `a_xor_b`.
+/// Further, he receives `hash_y1` and `hash_c_h3` from `H3`
+/// and `hash_c_h2` from `H2`.
+///
+/// ## Errors
+/// Propagates network errors. Further it returns an error when
+/// `hash_x1 != hash_y1` or `hash_c_h2 != hash_a_xor_b`
+/// or `hash_c_h3 != hash_a_xor_b`.
+async fn h1_verify<C: Context, S: BooleanArray>(
+    ctx: C,
+    keys: &[StdArray<Gf32Bit, 1>],
+    share_a_and_b: &[AdditiveShare<S>],
+    x1: &[S],
+) -> Result<(), Error> {
+    // compute hashes
+    // compute hash for x1
+    let hash_x1 = compute_row_hash::<S, _, _>(keys, x1);
+    // compute hash for A xor B
+    let hash_a_xor_b = compute_row_hash::<S, _, _>(
+        keys,
+        share_a_and_b
+            .iter()
+            .map(|share| share.left() + share.right()),
+    );
+
+    // setup channels
+    let h3_ctx = ctx
+        .narrow(&OPRFShuffleStep::HashesH3toH1)
+        .set_total_records(TotalRecords::specified(2)?);
+    let h2_ctx = ctx
+        .narrow(&OPRFShuffleStep::HashH2toH1)
+        .set_total_records(TotalRecords::specified(1)?);
```

Comment:
nit - just to save one extra question mark

```suggestion
        .set_total_records(TotalRecords::ONE);
```